connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -336,13 +332,16 @@ ClusterAssignment performTaskAssignment(
         log.debug("Incremental connector assignments: {}", incrementalConnectorAssignments);
         log.debug("Incremental task assignments: {}", incrementalTaskAssignments);
 
+        Map<String, Collection<String>> revokedConnectors = transformValues(toRevoke, ConnectorsAndTasks::connectors);
+        Map<String, Collection<ConnectorTaskId>> revokedTasks = transformValues(toRevoke, ConnectorsAndTasks::tasks);
+
         return new ClusterAssignment(
                 incrementalConnectorAssignments,
                 incrementalTaskAssignments,
-                transformValues(toRevoke, ConnectorsAndTasks::connectors),
-                transformValues(toRevoke, ConnectorsAndTasks::tasks),
-                connectorAssignments,
-                taskAssignments
+                revokedConnectors,
+                revokedTasks,
+                diff(connectorAssignments, revokedConnectors),
+                diff(taskAssignments, revokedTasks)
```

Comment:
Yep! There was a bug here. FWIW the same issue is already fixed by https://github.com/apache/kafka/pull/12019.

Reply:
I think the `connectorAssignments` should be equal to `diff(connectorAssignments, revokedConnectors)` and `taskAssignments == diff(taskAssignments, revokedTasks)`, because in L279:
https://github.com/apache/kafka/pull/11983/files#diff-e24067b121eb960feebfa099bd9c30382e330eaf5db39302a9d7a50e29b3acb4R279-R283

We already removed the revoked connectors/tasks. Is my understanding correct?

Reply:
I don't think so; it looks like we compute load-balancing revocations later on, around line 300. At line 279, the `completeWorkerAssignment` that we derive `connectorAssignments` and `taskAssignments` from only has the `deleted` connectors and tasks removed from it; everything else (including load-balancing revocations and duplicated assignments) is still included.